### PR TITLE
Add required build, Dependencies and github workflows

### PR DESCRIPTION
**Pull request by vimleshtna**

### CODE REVIEW

File: src/main/scala/uk/gov/nationalarchives/tdr/validation/MetadataCriteria.scala [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+    dependencies: Option[Map[String, List[MetadataCriteria]]] = None,
+    defaultValue: Option[String] = None
+)
+
+object MetadataProperty {
+  val closureType = "ClosureType"
+  val descriptiveType = "DescriptiveType"
+}
+
+object ErrorCode {
```
Comment (TomJKing) [inline src/main/scala/uk/gov/nationalarchives/tdr/validation/MetadataCriteria.scala:21]: `INVALID_DATE_FORMAT_ERROR` and `INVALID_NUMBER_ERROR` don't appear to be used any where in the code.

Is there some checks that are missing, or using the incorrect error, or are they simply not needed?

File: src/test/scala/uk/gov/nationalarchives/tdr/validation/DataTypeSpec.scala [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+
+    "checkValue should return an error if the value is not an integer" in {
+      Integer.checkValue("e", criteria) should be(Some(NUMBER_ONLY_ERROR))
+    }
+
+    "checkValue should return an error if the value is a negative number" in {
+      Integer.checkValue("-1", criteria) should be(Some(NEGATIVE_NUMBER_ERROR))
+    }
+
+    "checkValue should not return any error if the value is a valid number" in {
```
Comment (TomJKing) [inline src/test/scala/uk/gov/nationalarchives/tdr/validation/DataTypeSpec.scala:25]: Spelling `error` should be `errors`

File: src/test/scala/uk/gov/nationalarchives/tdr/validation/DataTypeSpec.scala [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+  "DateTime" should {
+
+    val criteria = MetadataCriteria("Property1", DateTime, true, false, false, Nil, None, None)
+
+    "checkValue should not return amy error if the date is valid" in {
+      DateTime.checkValue("1990-12-10T00:00:00", criteria) should be(None)
+      DateTime.checkValue("2000-2-29T00:00:00", criteria) should be(None)
+    }
+
+    "checkValue should not return amy error if the value is empty but it is not mandatory" in {
```
Comment (TomJKing) [inline src/test/scala/uk/gov/nationalarchives/tdr/validation/DataTypeSpec.scala:39]: Spelling, `amy` should be `any` and `error` should be `errors`.

File: src/test/scala/uk/gov/nationalarchives/tdr/validation/DataTypeSpec.scala [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+      DateTime.checkValue("", criteria.copy(required = false)) should be(None)
+    }
+
+    "checkValue should not return amy error if future date is allowed" in {
```
Comment (TomJKing) [inline src/test/scala/uk/gov/nationalarchives/tdr/validation/DataTypeSpec.scala:43]: Spelling, `amy` should be `any` and `error` should be `errors`

File: src/test/scala/uk/gov/nationalarchives/tdr/validation/DataTypeSpec.scala [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+    "checkValue should return an error if future date is not allowed" in {
+      DateTime.checkValue("2050-2-1T00:00:00", criteria) should be(Some(FUTURE_DATE_ERROR))
+    }
+  }
+
+  "Text" should {
+
+    val criteria = MetadataCriteria("Property1", Text, true, false, false, Nil, None, None)
+
+    "checkValue should not return any error if the value is valid" in {
```
Comment (TomJKing) [inline src/test/scala/uk/gov/nationalarchives/tdr/validation/DataTypeSpec.scala:83]: Spelling `error` should be `errors`

File: src/test/scala/uk/gov/nationalarchives/tdr/validation/DataTypeSpec.scala [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+      Text.checkValue("hello", criteria) should be(None)
+    }
+
+    "checkValue should not return any error if the value is empty but it is not mandatory" in {
```
Comment (TomJKing) [inline src/test/scala/uk/gov/nationalarchives/tdr/validation/DataTypeSpec.scala:87]: Spelling `error` should be `errors`

File: src/test/scala/uk/gov/nationalarchives/tdr/validation/DataTypeSpec.scala [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+      Text.checkValue("", criteria.copy(required = false)) should be(None)
+    }
+
+    "checkValue should return an error if the value is empty" in {
```
Comment (TomJKing) [inline src/test/scala/uk/gov/nationalarchives/tdr/validation/DataTypeSpec.scala:91]: Reword test slightly to: `... value is empty and mandatory"`

File: src/test/scala/uk/gov/nationalarchives/tdr/validation/DataTypeSpec.scala [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+      Text.checkValue("", criteria) should be(Some(EMPTY_VALUE_ERROR))
+    }
+
+    "checkValue should not return any error if the value has multiple values but multiple values are allowed" in {
```
Comment (TomJKing) [inline src/test/scala/uk/gov/nationalarchives/tdr/validation/DataTypeSpec.scala:95]: Spelling `error` should be `errors`

File: src/test/scala/uk/gov/nationalarchives/tdr/validation/DataTypeSpec.scala [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+      Text.checkValue("22,44", criteria.copy(definedValues = List("22", "44"), isMultiValueAllowed = true)) should be(None)
+    }
+
+    "checkValue should return an error if the value has multiple value but multiple values are not allowed" in {
```
Comment (TomJKing) [inline src/test/scala/uk/gov/nationalarchives/tdr/validation/DataTypeSpec.scala:99]: Spelling `value` should be `values`

File: src/test/scala/uk/gov/nationalarchives/tdr/validation/DataTypeSpec.scala [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+      Text.checkValue("22,44", criteria.copy(definedValues = List("22", "44"))) should be(Some(MULTI_VALUE_ERROR))
+    }
+
+    "checkValue should return an error if the value is not matching with defined values" in {
```
Comment (TomJKing) [inline src/test/scala/uk/gov/nationalarchives/tdr/validation/DataTypeSpec.scala:103]: Reword test slightly: `... is not matching its defined values"`

File: src/test/scala/uk/gov/nationalarchives/tdr/validation/DataTypeSpec.scala [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+    }
+
+    "checkValue should return an error if the value is not matching with the defined values" in {
+      Boolean.checkValue("true", criteria, None) should be(Some(UNDEFINED_VALUE_ERROR))
+    }
+
+    "checkValue should return an error if the required property value is empty" in {
+      Boolean.checkValue("yes", criteria.copy(requiredProperty = Some("property2")), Some(Metadata("property2", ""))) should be(Some(REQUIRED_PROPERTY_IS_EMPTY))
+    }
+
```
Comment (TomJKing) [inline src/test/scala/uk/gov/nationalarchives/tdr/validation/DataTypeSpec.scala:135]: Unnecessary line break

File: src/test/scala/uk/gov/nationalarchives/tdr/validation/DataTypeSpec.scala [hunk 12]
```diff
@@ -0,0 +1,137 @@
...
+    "checkValue should return an error if the value is not matching with defined values" in {
+      Text.checkValue("22,44", criteria.copy(definedValues = List("22", "33"), isMultiValueAllowed = true)) should be(Some(UNDEFINED_VALUE_ERROR))
+    }
+  }
+
+  "Boolean" should {
+
+    val criteria = MetadataCriteria("Property1", Boolean, true, false, false, List("yes", "no"), None, None)
+
+    "checkValue should not return any error if the value is valid" in {
```
Comment (TomJKing) [inline src/test/scala/uk/gov/nationalarchives/tdr/validation/DataTypeSpec.scala:112]: Spelling `error` should be `errors`

File: src/test/scala/uk/gov/nationalarchives/tdr/validation/DataTypeSpec.scala [hunk 13]
```diff
@@ -0,0 +1,137 @@
...
+      Boolean.checkValue("yes", criteria, None) should be(None)
+    }
+
+    "checkValue should not return any error if the value is empty but required property is empty too" in {
```
Comment (TomJKing) [inline src/test/scala/uk/gov/nationalarchives/tdr/validation/DataTypeSpec.scala:116]: Spelling `error` should be `errors`

File: src/test/scala/uk/gov/nationalarchives/tdr/validation/DataTypeSpec.scala [hunk 14]
```diff
@@ -0,0 +1,137 @@
...
+
+    "checkValue should return an error if the value is empty and no required property" in {
+      Boolean.checkValue("", criteria, None) should be(Some(NO_OPTION_SELECTED_ERROR))
+    }
+
+    "checkValue should return an error if the value is empty and has required property" in {
+      Boolean.checkValue("", criteria.copy(requiredProperty = Some("property2")), Some(Metadata("property2", "value"))) should be(Some(NO_OPTION_SELECTED_ERROR))
+    }
+
+    "checkValue should return an error if the value is not matching with the defined values" in {
```
Comment (TomJKing) [inline src/test/scala/uk/gov/nationalarchives/tdr/validation/DataTypeSpec.scala:128]: Reword test slightly: `... value does not match its defined values"`

File: src/test/scala/uk/gov/nationalarchives/tdr/validation/MetadataValidationSpec.scala [hunk 15]
```diff
@@ -0,0 +1,231 @@
+package uk.gov.nationalarchives.tdr.validation
+
+import org.scalatest.flatspec.AnyFlatSpec
+import org.scalatest.matchers.should.Matchers._
+import uk.gov.nationalarchives.tdr.validation.ErrorCode._
+import uk.gov.nationalarchives.tdr.validation.MetadataProperty.closureType
+
+class MetadataValidationSpec extends AnyFlatSpec {
+
+  "validateClosureMetadata" should "validate closure metadata with valid values" in {
```
Comment (TomJKing) [inline src/test/scala/uk/gov/nationalarchives/tdr/validation/MetadataValidationSpec.scala:10]: Reword test slightly: `... validate closure metadata"`

File: src/test/scala/uk/gov/nationalarchives/tdr/validation/MetadataValidationSpec.scala [hunk 16]
```diff
@@ -0,0 +1,231 @@
...
+    val closureMetadataCriteria = MetadataCriteria("ClosureType", Boolean, true, false, false, List("yes", "no"), None, Some(Map("Closed" -> dependentMetadataCriteria)), None)
+    val input = List(
+      Metadata("Property1", "yes")
+    )
+    val metadataValidation = new MetadataValidation(closureMetadataCriteria, Nil)
+    val error = metadataValidation.validateClosureMetadata(input)
+    error should be(List(Error(closureType, CLOSURE_STATUS_IS_MISSING)))
+  }
+
+  "validateClosureMetadata" should "not return any error if closure status is open and closure metadata has empty or default value" in {
```
Comment (TomJKing) [inline src/test/scala/uk/gov/nationalarchives/tdr/validation/MetadataValidationSpec.scala:60]: Reword test slightly: `... any errors ... has empty or default values"`

File: src/test/scala/uk/gov/nationalarchives/tdr/validation/MetadataValidationSpec.scala [hunk 17]
```diff
@@ -0,0 +1,231 @@
...
+    error should be(
+      List(
+        Error("Property1", EMPTY_VALUE_ERROR),
+        Error("Property2", EMPTY_VALUE_ERROR),
+        Error("Property4", EMPTY_VALUE_ERROR)
+      )
+    )
+  }
+
+  "validateClosureMetadata" should "not return an error if a closure metadata dependent by a descriptive metadata which is empty" in {
```
Comment (TomJKing) [inline src/test/scala/uk/gov/nationalarchives/tdr/validation/MetadataValidationSpec.scala:156]: Reword test slightly: `... if closure metadata is dependent on descriptive metadata which is empty"`

File: src/test/scala/uk/gov/nationalarchives/tdr/validation/MetadataValidationSpec.scala [hunk 18]
```diff
@@ -0,0 +1,231 @@
...
+      Metadata("Property1", ""),
+      Metadata("Property11", ""),
+      Metadata("Property5", "")
+    )
+    val metadataValidation = new MetadataValidation(closureMetadataCriteria, Nil)
+    val error = metadataValidation.validateClosureMetadata(input)
+    error should be(Nil)
+  }
+
+  "validateDescriptiveMetadata" should "validate descriptive metadata with valid values" in {
```
Comment (TomJKing) [inline src/test/scala/uk/gov/nationalarchives/tdr/validation/MetadataValidationSpec.scala:182]: Reword test slightly: `"validate descriptive metadata"`

File: src/test/scala/uk/gov/nationalarchives/tdr/validation/MetadataValidationSpec.scala [hunk 19]
```diff
@@ -0,0 +1,231 @@
...
+      Metadata("Property1", ""),
+      Metadata("Property2", ""),
+      Metadata("Property3", "test")
+    )
+    val metadataValidation = new MetadataValidation(closureMetadataCriteria, descriptiveMetadataCriteria)
+    val error = metadataValidation.validateDescriptiveMetadata(input)
+    error should be(Nil)
+  }
+
+  "validateMetadata" should "validate closure & descriptive metadata" in {
```
Comment (TomJKing) [inline src/test/scala/uk/gov/nationalarchives/tdr/validation/MetadataValidationSpec.scala:200]: Reword test slightly: `... closure and description metadata"`

File: src/main/scala/uk/gov/nationalarchives/tdr/validation/DataType.scala [hunk 20]
```diff
@@ -0,0 +1,157 @@
...
+    value match {
+      case "" if criteria.required            => Some(EMPTY_VALUE_ERROR)
+      case t if allCatch.opt(t.toInt).isEmpty => Some(NUMBER_ONLY_ERROR)
+      case t if t.toInt < 0                   => Some(NEGATIVE_NUMBER_ERROR)
+      case _                                  => None
+    }
+  }
+}
+
+case object DateTime extends AnyRef with DataType with Product with Serializable {
```
[the remark this answers is not in the document]
Comment (vimleshtna) [reply src/main/scala/uk/gov/nationalarchives/tdr/validation/DataType.scala]: Actually, we are showing specific errors for day/month/year which can not be possible with the API.

File: src/main/scala/uk/gov/nationalarchives/tdr/validation/DataType.scala [hunk 21]
```diff
@@ -0,0 +1,157 @@
+package uk.gov.nationalarchives.tdr.validation
+
+import uk.gov.nationalarchives.tdr.validation.ErrorCode._
+
+import java.time.{LocalDateTime, Year}
+import scala.util.control.Exception.allCatch
+
+sealed trait DataType extends AnyRef
+
+case object Integer extends AnyRef with DataType with Product with Serializable {
```
[the remark this answers is not in the document]
Comment (vimleshtna) [reply src/main/scala/uk/gov/nationalarchives/tdr/validation/DataType.scala]: Yes, it's not required, It's a copy-paste error. I have removed it.

File: src/main/scala/uk/gov/nationalarchives/tdr/validation/MetadataCriteria.scala (hunk 1 above)
Comment (vimleshtna) [reply src/main/scala/uk/gov/nationalarchives/tdr/validation/MetadataCriteria.scala:21]: Good catch. Added a check for `INVALID_DATE_FORMAT_ERROR` and `INVALID_NUMBER_ERROR` is not needed so I removed it

**Comment by vimleshtna** [comment]
> set up, so these will need to be added to the

I have already added all the secrets https://github.com/nationalarchives/tdr-terraform-github/pull/10/files

**Review by TomJKing — APPROVED** [review]
Looks good. Thanks.

**Pull request merged**